public/pages/DetectorConfig/containers/Features.tsx

```diff
@@ -93,6 +97,7 @@ export class Features extends Component<FeaturesProps, FeaturesState> {
 
   public render() {
     const featureAttributes = get(this.props.detector, 'featureAttributes', []);
+    const shingleSize = get(this.props.detector, 'shingleSize', SHINGLE_SIZE);
```

Comment:
Is `shingleSize` optional for backend Java Detector class? If yes, suggest to set `shingleSize=8` for java detector instance. 
So if we change the default SHINGLE_SIZE in future, we don't need to worry about the backward compatibility about old detectors. For example, user create detector with default SHINGLE_SIZE 8 now, later we change the default SHINGLE_SIZE as 4. If we don't set this detector's shingle size as 8 in data model, we may reset it as 4, and model will fail as checkpoint can't match current detector shingle size configuration.

Reply:
Could you clarify what change you're suggesting in the frontend code? Are you suggesting the SHINGLE_SIZE constant be removed from the frontend?

The `shingleSize` parameter is optional in the API, and the backend already defaults the shingle size to 8 if the parameter is not provided in the request. However, this change will only be applied to detectors which get created or updated after this backend change was made.

The reason the default shingle size is also specified in the frontend is precisely for backwards-compatibility. Detectors that were created prior to the recent backend change and have not been updated since then will not have a "shingleSize" field in the stored detector configuration, and it would not be a good user experience to see a "null" or blank in the Window Size field when they view their model configurations in the UI. Even if there is no "shingleSize" field stored on the detector configuration, the user should be informed by the UI that a shingle size of 8 is being used for that detector. In addition, the shingle size needs to be known by the frontend to calculate detector initialization time (see `isDetectorInitOverTime`). (The SHINGLE_SIZE constant was actually already existing in the code for this reason. This constant did not get added in this change.)

If they save any changes on the model configuration page for the existing detector, even if they didn't change the window size, going forward the "shingleSize" field will get stored on the detector configuration for the future, but until all existing detectors have been updated to have that field, the default value is needed to ensure the model configuration display the correct value and calculates the correct detector initialization time.

Reply:
As I said "suggest to set shingleSize=8 for java detector instance." now you have already set it as 8, "the backend already defaults the shingle size to 8 if the parameter is not provided in the request". So that will be ok.